### DISCUSSION TITLE
AAeon Jeston Xavier NX problem with 5.10 update

### DISCUSSION THREAD

**Discussion by EtimosGH** [OE4T/meta-tegra#1289] [Q&A]
Hello everyone,

I'm working on the update from 4.9 to 5.10 on multiple devices. I work with a Jetson Xavier AGX Dev Kit, Jetson Orin AGX Dev kit and a Jetson Xavier NX on a carrier board from aaeon (Boxer 8253AI). I have no problem with Xavier and Orin AGX, and I have no problem building and flashing. However, when I build with last kirkstone/langdale (tried both) and flash the Xavier NX, it doesn't show anything on the screen. I don't use an SD card/nvme/ssd. Just basic emmc. I flash using the `doflash.sh` script.
I tried to build basic images from tegra-demo-distro, demo-image-egl and demo-image-sato, with no luck. However, the jetpack from Nvidia does work.

The thing is, I don't have the required cable to read from serial console, which makes debugging a bit hard (It is not on a usb output, but on what looks like a JST connector on the board, [chapter 2.4.9 on this documentation](https://newdata.aaeon.com.tw/DOWNLOAD/MANUAL/BOXER-8253AI%20Manual%207th%20Ed.pdf) ). I did manage to connect to RS-232 but the only output I managed to get is after flashing with Jetpack I have this : 
```
**  WARNING: Test Key is used.  **
L4TLauncher: Attempting Direct Boot
EFI stub: Booting Linux Kernel...
EFI stub: Using DTB from configuration table
EFI stub: Loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path
EFI stub: Exiting boot services and installing virtual address map...
```

I also tried to use the `initrd-flash` script, but it seems to stop here:
```
Starting at 2023-07-05T16:49:36+02:00                                                                                                                                                                              
Machine:       aaeon-boxer-8253ai                                                                                                                                                                                  
Rootfs device: mmcblk0p1                                                                                                                                                                                           
Found Jetson device in recovery mode at USB 1-4                                                                                                                                                                    
== Step 1: Signing binaries at 2023-07-05T16:49:36+02:00 ==                                                                                                                                                        
== Step 2: Boot Jetson via RCM at 2023-07-05T16:49:56+02:00 ==                                                                                                                                                     
Found Jetson device in recovery mode at USB 1-4                                                                                                                                                                    
== Step 3: Sending flash sequence commands at 2023-07-05T16:50:27+02:00 ==                                                                                                                                         
Waiting for USB storage device flashpkg from 0d0c0009................................................
```

I'm open to any suggestion on what to try next.

Thanks in advance for the help, and also thanks for meta-tegra, it's awesome.

**Comment by madisongh** [comment]
Does AAeon provide an adaptation of JetPack 5/L4T R35.x for that carrier?  If they do, try booting that on the device, and if it works, try a side-by-side comparison of the boot- and kernel-related files that it uses against the ones generated out of your Yocto build.  If not, how did you construct the device tree and other boot-time configuration files?

**Reply by EtimosGH** [reply]
> Does AAeon provide an adaptation of JetPack 5/L4T R35.x for that carrier?

They mention a AAEON ACLinux4.9 on this particular board, but no 5.0 adaptation. I already made a tech support ticket to ask for this update, but they do take some time to respond (last time I asked them something it took about 2 months or so).

> if not, how did you construct the device tree and other boot-time configuration files?

For now, with version 4.9, I'm using a dtb we extracted from their ACLinux 4.9 and a machine conf for adding some extra kernel modules and firmware. I override the dtb with a `linux-tegra_5.10.bbappend` like this 
```
SRC_URI:append:aaeon-boxer-8253ai = " file://aaeon-boxer-8253ai/tegra194-p3668-all-p3509-0000-boxer8253.dtb"
do_install:append:aaeon-boxer-8253ai() {
    install -m 0644 ${WORKDIR}/aaeon-boxer-8253ai/tegra194-p3668-all-p3509-0000-boxer8253.dtb ${D}/${KERNEL_IMAGEDEST}/tegra194-p3668-all-p3509-0000.dtb
}
```
I also override three variables, which seemed to induce some latency during booting : 
```
CONFIG_ACS=y
CONFIG_FW_LOADER_USER_HELPER=n
CONFIG_FW_LOADER_USER_HELPER_FALLBACK=n
```

Machine conf looks like this:
```
MACHINEOVERRIDES = "jetson-xavier-nx-devkit-emmc:${MACHINE}"
require conf/machine/jetson-xavier-nx-devkit-emmc.conf
PACKAGE_EXTRA_ARCHS:append = " jetson-xavier-nx-devkit-emmc"

#KERNEL_DEVICETREE = "tegra194-p3668-all-p3509-0000-boxer8253.dtb"

MACHINE_FEATURES += "wifi bluetooth"
MACHINE_EXTRA_RRECOMMENDS += "kernel-module-iwlmvm kernel-module-iwlwifi linux-firmware-iwlwifi-8265"
MACHINE_EXTRA_RRECOMMENDS += "kernel-module-btusb linux-firmware-ibt"
TEGRA_BT_SUPPORT_PACKAGE = ""

NVPMODEL_CONFIG_DEFAULT = "2"
``` 

With 4.9 we got it working without issue, but with 5.10 only Nvidia's Jetpack works. 
I'll call Aaeon to ask if they can provide a UART cable for their strange port or try to tinker a cable for it, so I can have access to the serial console.  
Do you think I can take advantage of the fact that Jetpack 5.10 works to create an updated dtb ? I don't have a lot of experience doing that sort of things, so I don't really now if that's possible.

**Reply by madisongh** [reply]
If you can get NVIDIA's distro to boot on the device with JetPack, you should be able to get something bootable out of a Yocto build, if you can match up the device tree and the boot-time config files.  Getting a serial console attached to see what's actually happening on the device while it boots would help with figuring that out.

**Reply by EtimosGH** [reply]
I asked Aaeon for a cable for the debug console, for now I'm waiting for a response on their side. I'll get back when I (hopefully) have the said cable. 
For the device tree, I'll try to decompile the dtb and compare the dts it with the one given we get with nvidia, but for now I'll wait for the cable to be sure it's the problem I'm facing. Thanks for the help ! I'll be back (let's hope not too far from now) with either some good news, or with a nice serial console log.
